runtime/gc_vlhgc/MemorySubSpaceTarok.cpp

```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+	 * For each % above heapExpansionGCTimeThreshold, expand more aggresively. 
+	 * Ie, if heapExpansionGCTimeThreshold = 13, and measured hybrid overhead is 20, we want to expand by a larger amount than if measured hybrid overhead is 14.
+	 */
+	if (!foundAcceptableHeapSizeChange) {
+		/* Expansion and contraction may increase/decrease by larger/smaller amounts */
+		uintptr_t sizeChangeFactor = 0;
+
+		/* percentDiff is represented as percent between 0 - 100 */
+		double percentDiff = 0.0;
+
+		if (currentHybridHeapScore >= (double)_extensions->heapExpansionGCTimeThreshold) {
+			/* Try to bring hybridHeapScore a little bit below _extensions->heapExpansionGCTimeThreshold */
+			percentDiff = currentHybridHeapScore - (double)_extensions->heapExpansionGCTimeThreshold;
+
+			/* Limit the percent difference to prevent any accidental big changes. 
+			 * This helps deal with cases with one GC that is too long for whatever reason, causing a massive, undesired increase in heap size 
+			 */
+			percentDiff = OMR_MAX(percentDiff, 5.0);
+
+			/* Be a bit more aggressive with expansion than contraction.  */	
+			sizeChangeFactor = 2;	
+
+		} else if (currentHybridHeapScore <= (double)_extensions->heapContractionGCTimeThreshold) {
+			/* Try to bring hybridHeapScore a little bit above _extensions->heapContractionGCTimeThreshold.
+			 * If this doesn't cause the hybrid heap score to increase enough so that it is within acceptable bounds, this path will be taken 
+			 * again later, and another small contraction will occur.
+			 */
+			percentDiff = currentHybridHeapScore - (double)_extensions->heapContractionGCTimeThreshold;
+			sizeChangeFactor = 1;
+		} 
+
+		/* Since percentDiff is 0 - 100, make sure to convert it to 0-1 percentage */
+		double heapSizePercentChange = (1.0 + ((double)sizeChangeFactor * (percentDiff / 100 )));
+		Trc_MM_MemorySubSpaceTarok_getHeapSizeWithinBounds_2(env->getLanguageVMThread(), heapSizePercentChange);
+		recommendedHeapSize = (uintptr_t)(heapSizePercentChange * recommendedHeapSize);
+	}
+
+	return recommendedHeapSize;
+}
+
+
+double 
+MM_MemorySubSpaceTarok::calculateGcPctForHeapChange(MM_EnvironmentBase *env, intptr_t heapSizeChange)
+{
+	MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
+
+	/* 
+	 * Since we only attempt to resize after PGC's and Global GC's, if the PGC count has increased, it means we are resizing after a PGC
+	 * and the GMP and PGC statistics we have collected, are useable. Otherwise, use backup methods which are independant of heapSizeChange
+	 */
+
+	if (_previouslyObservedPGCCount != envVLHGC->_heapSizingData.pgcCountSinceGMPEnd) {
+		/* Resizing after a PGC */
+
+		uintptr_t pgcCount = envVLHGC->getRepresentativePgcPerGmpCount();
+
+		if ((0 == pgcCount) && (0 == _lastObservedGcPercentage)) {
+			/* Very first time we are resizing, assume GC % is heapExpansionGCTimeThreshold */
+			_lastObservedGcPercentage = (double)_extensions->heapExpansionGCTimeThreshold;
```

Comment:
would  (expansion + contraction)/2 be more reasonable?

Reply:
My thought behind making this `heapExpansionGCTimeThreshold` is so that heap expansion would be "easier" to suggest at the very beginning of the run, although it likely doesn't matter too much - If free memory is very tight, the heuristic will suggest expansion anyways. Happy to change it to `(expansion + contraction)/2` if you prefer

Reply:
I leave it up to you - probably not much difference.

Reply:
I'll keep it as `heapExpansionGCTimeThreshold `, but I will elaborate on the comment